aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move

```diff
@@ -1,97 +1,160 @@
 /// This module provides an interface for aggregators (version 2).
-/// Only skeleton - for AggregagtorSnapshot - is provided at this time,
-/// to allow transition of usages.
 module aptos_framework::aggregator_v2 {
+    use std::error;
     use std::string::String;
 
+    /// The value of aggregator overflows. Raised by uncoditional add() call
+    const EAGGREGATOR_OVERFLOW: u64 = 1;
+
+    /// The value of aggregator underflows (goes below zero). Raised by uncoditional sub() call
+    const EAGGREGATOR_UNDERFLOW: u64 = 2;
+
     /// The generic type supplied to the aggregator snapshot is not supported.
     const EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE: u64 = 5;
 
     /// The aggregator snapshots feature flag is not enabled.
     const EAGGREGATOR_SNAPSHOTS_NOT_ENABLED: u64 = 6;
 
+    /// The generic type supplied to the aggregator is not supported.
+    const EUNSUPPORTED_AGGREGATOR_TYPE: u64 = 7;
+
+    /// Represents an integer which supports parallel additions and subtractions
+    /// across multiple transactions. See the module description for more details.
+    ///
+    /// Currently supported types for Element are u64 and u128.
+    struct Aggregator<IntElement> has store {
+        value: IntElement,
+        max_value: IntElement,
+    }
+
     struct AggregatorSnapshot<Element> has store, drop {
         value: Element,
     }
 
+    /// Returns `max_value` exceeding which aggregator overflows.
+    public fun max_value<IntElement: copy + drop>(aggregator: &Aggregator<IntElement>): IntElement {
+        aggregator.max_value
+    }
+
+    /// Creates new aggregator, with given 'max_value'.
+    ///
+    /// Currently supported types for Element are u64 and u128.
+    /// EAGGREGATOR_ELEMENT_TYPE_NOT_SUPPORTED raised if called with a different type.
+    public native fun create_aggregator<IntElement: copy + drop>(max_value: IntElement): Aggregator<IntElement>;
+
+    /// Creates new aggregator, without any 'max_value' on top of the implicit bound restriction
+    /// due to the width of the type (i.e. MAX_U64 for u64, MAX_U128 for u128).
+    ///
+    /// Currently supported types for Element are u64 and u128.
+    /// EAGGREGATOR_ELEMENT_TYPE_NOT_SUPPORTED raised if called with a different type.
+    public native fun create_unbounded_aggregator<IntElement: copy + drop>(): Aggregator<IntElement>;
```

Comment:
Does this have to be a native function? It can simply call create_aggregator with an appropriate max_value right?

Reply:
I don't think it can be done, because the type is generic.

i.e. something like this doesn't compile:

    public fun create_unbounded_aggregator<IntElement: copy + drop>(): Aggregator<IntElement> {
        if (type_info::type_of<IntElement>() == type_info::type_of<u64>()) {
            create_aggregator(MAX_U64)
        } else if (type_info::type_of<IntElement>() == type_info::type_of<u128>()) {
            create_aggregator(MAX_U128)
        } else {
            abort error::invalid_argument(EUNSUPPORTED_AGGREGATOR_TYPE);
        }
    }


this compiles:
`if (type_info::type_of<Element>() == type_info::type_of<u64>()) {`

this doesn't:
`return create_aggregator(MAX_U64)`
when return type of the function is : Aggregator<IntElement>

Compiler could probably infer something, but even in C++/rust not sure that would pass. Not sure if there is a different way to write this.